system/lib/pthread/proxying.c

```diff
@@ -620,12 +673,18 @@ double _emscripten_run_on_main_thread_js(int func_index,
   em_proxying_queue* q = emscripten_proxy_get_system_queue();
   pthread_t target = emscripten_main_runtime_thread_id();
 
-  if (sync) {
+  if (sync && !promise) {
     if (!emscripten_proxy_sync(q, target, run_js_func, &f)) {
       assert(false && "emscripten_proxy_sync failed");
       return 0;
     }
     return f.result;
+  } else if (sync && promise) {
+    if (!emscripten_proxy_promise_await(q, target, run_js_promise_await, &f)) {
+      assert(false && "emscripten_proxy_promise failed");
+      return 0;
+    }
+    return f.result;
   }
```

Comment:
Maybe structure this as:

```
if (sync) {
   if (promise) {
   } else { 
   }
   return f.result;
}
```

Also the fact that we have `int sync` and `int promise` here as a separate arguments, but it doesn't make sense for `promise` to be set without `sync` seem a little odd.

At the very least we should probably assert that promise is never true when sync is not.